lib/fswatcher/fswatcher.go

```diff
@@ -0,0 +1,577 @@
+		}
+		w.resetNotifyTimerIfNeeded()
+		return
+	}
+
+	parentDir := w.rootEventDir
+
+	// Check if any parent directory is already tracked or will exceed
+	// events per directory limit bottom up
+	pathSegments := strings.Split(filepath.ToSlash(evPath), "/")
+
+	// As root dir cannot be further aggregated, allow up to maxFiles
+	// children.
+	localMaxFilesPerDir := maxFiles
+	var currPath string
+	for i, pathSegment := range pathSegments[:len(pathSegments)-1] {
+		currPath = filepath.Join(currPath, pathSegment)
+
+		if event, ok := parentDir.events[currPath]; ok {
+			event.lastModTime = evTime
+			event.evType |= evType
+			l.Debugf("%v Parent %s (type %s) already tracked: %s", w, currPath, event.evType, evPath)
+			return
+		}
+
+		if parentDir.childCount() == localMaxFilesPerDir {
+			l.Debugf("%v Parent dir %s already has %d children, tracking it instead: %s", w, currPath, localMaxFilesPerDir, evPath)
+			w.aggregateEvent(filepath.Dir(currPath),
+				evTime, evType)
+			return
+		}
+
+		// If there are no events below path, but we need to recurse
+		// into that path, create eventDir at path.
+		if _, ok := parentDir.dirs[currPath]; !ok {
+			l.Debugf("%v Creating eventDir: %s", w, currPath)
+			parentDir.dirs[currPath] = newEventDir(currPath, parentDir)
+		}
+		parentDir = parentDir.dirs[currPath]
+
+		// Reset allowed children count to maxFilesPerDir for non-root
+		if i == 0 {
+			localMaxFilesPerDir = maxFilesPerDir
+		}
+	}
+
+	if event, ok := parentDir.events[evPath]; ok {
+		event.lastModTime = evTime
+		event.evType |= evType
+		l.Debugf("%v Already tracked (type %v): %s", w, event.evType, evPath)
+		return
+	}
+
+	childDir, ok := parentDir.dirs[evPath]
+
+	// If a dir existed at path, it would be removed from dirs, thus
+	// childCount would not increase.
+	if !ok && parentDir.childCount() == localMaxFilesPerDir {
+		l.Debugf("%v Parent dir already has %d children, tracking it instead: %s", w, localMaxFilesPerDir, evPath)
+		w.aggregateEvent(filepath.Dir(evPath), evTime, evType)
+		return
+	}
+
+	firstModTime := evTime
+	if ok {
+		firstModTime = childDir.firstModTime()
+		evType |= childDir.eventType()
+		delete(parentDir.dirs, evPath)
+	}
+	l.Debugf("%v Tracking (type %v): %s", w, evType, evPath)
+	parentDir.events[evPath] = &event{
+		path:         evPath,
+		firstModTime: firstModTime,
+		lastModTime:  evTime,
+		evType:       evType,
+	}
+	w.resetNotifyTimerIfNeeded()
+}
+
+func (w *watcher) actOnTimer() {
+	eventCount := w.rootEventDir.eventCount()
+	if eventCount == 0 {
+		l.Debugln(w, "No tracked events, waiting for new event.")
+		w.notifyTimerNeedsReset = true
+		return
+	}
+	oldFsEvents := w.popOldEvents(w.rootEventDir, time.Now())
+	if len(oldFsEvents) == 0 {
+		l.Debugln(w, "No old fs events")
+		w.resetNotifyTimer(w.notifyDelay)
+		return
+	}
+	// Sending to channel might block for a long time, but we need to keep
+	// reading from notify backend channel to avoid overflow
+	go w.notify(oldFsEvents)
```

Comment:
We should somehow throttle this, as this can spawn stuff indefinitely back to back if there are a lot of changes happening.

I'd be nice to track if the last scan finished, and if it hasn't keep aggregating.